mono/metadata/threads.c

```diff
@@ -6583,6 +6582,353 @@ summarizer_supervisor_end (SummarizerSupervisorState *state)
+		STATE_CASE(LEADER_STATE_WAIT_BEFORE_SUSPEND_OTHERS);
+		STATE_CASE(LEADER_STATE_SUSPEND_OTHERS);
+		STATE_CASE(LEADER_STATE_WAIT_FOR_ORIGINATOR_STACK);
+		STATE_CASE(LEADER_STATE_SUMMARIZER_STATE_TERM);
+#undef STATE_CASE
+	default:
+		return "<unknown leader state>";
+	}
+}
+
+/* Called by the leader to wait for a command from the crash originator */
+static gboolean
+summarizer_leader_wait_for_command (int *leader_command, int *leader_state)
+{
+	MONO_ENTER_GC_SAFE;
+	/* allow interruptions */
+	while (mono_os_sem_wait (&summarizer_leader_data.leader_commanded, MONO_SEM_FLAGS_ALERTABLE) < 0);
+	MONO_EXIT_GC_SAFE;
+	mono_memory_barrier ();
+	*leader_command = summarizer_leader_data.leader_command;
+	if (*leader_command == LEADER_COMMAND_CANCEL) {
+		*leader_state = LEADER_STATE_READY;
+		return FALSE;
+	}
+	return TRUE;
+}
+
+
+/* Called by the originator to post commands to the leader.  Usually just to proceed to the next state */
+static void
+summarizer_leader_post_command (int command)
+{
+	summarizer_leader_data.leader_command = command;
+	mono_memory_barrier ();
+	mono_os_sem_post (&summarizer_leader_data.leader_commanded);
+}
+
+static void
+summarizer_leader_collect_thread_ids (SummarizerGlobalState *state);
+static void
+summarizer_leader_suspend_others (SummarizerGlobalState *state, MonoNativeThreadId originator, int originator_idx);
+static void
+summarizer_leader_set_originator_summary (MonoThreadSummary *orignator_summary);
+static void
+summarizer_state_get_index_for_thread (SummarizerGlobalState *state, MonoNativeThreadId current, int *my_index);
+static void
+summarizer_state_wait_and_term (MonoNativeThreadId caller_tid, SummarizerGlobalState *state, gchar **out, gchar *working_mem, size_t provided_size, MonoThreadSummary *originator_summary);
+static void
+summarizer_leader_adjust_tids_for_foreign_originator (void);
+
+static void
+summarizer_leader (void)
+{
+	MonoInternalThread *thread = mono_thread_internal_current ();
+	thread->flags |= MONO_THREAD_FLAG_DONT_MANAGE;
+
+	/*
+	 * This thread must not be stopped by the profiler or the STW
+	 * machinery. While collecting crashes it also violates the coop GC
+	 * rules by accessing managed memory to gather crash reports.  But
+	 * crash reporting has its own signal-based mechanism to interrupt the
+	 * other threads, so this is okay.
+	 */
+	mono_thread_info_set_flags (MONO_THREAD_INFO_FLAGS_NO_GC | MONO_THREAD_INFO_FLAGS_NO_SAMPLE);
+
+	mono_thread_set_name_constant_ignore_error (thread, "Crash Report Leader", MonoSetThreadNameFlag_None);
+
+	mono_thread_set_state (mono_thread_internal_current (), ThreadState_Background);
+
+	/* This thread is always in async context */
+	mono_thread_info_set_is_async_context (TRUE);
+	
+	mono_atomic_store_i32 (&summarizer_leader_data.leader_running, 1);
+	int leader_state = LEADER_STATE_READY;
+	while (TRUE) {
+		LEADER_LOG ("leader in state %s\n", leader_state_name (leader_state));
+		switch (leader_state) {
```

Comment:
I find the state machine to be over engineering for this use case. The transitions are just from one state to the next, and going back to original ready when something bad happens. This could be easily implemented with a simple loop.

Reply:
Yea, I added it because I wasn't sure how complicated the cleanup might have to be. But in the end it's just straight line code with pauses and one early exit. I'll remove it.